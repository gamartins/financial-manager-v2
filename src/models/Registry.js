export default class Registry {
  static TYPE = {
    PROVENTO: 0,
    DESCONTO: 1,
  }

  static STATUS = {
    NOT_PAID: 0,
    PAID: 1,
    PAID_FINANCIAL_CASH: 2,
    RESERVERD_SAVINGS: 3,
  }

  static getStatusString(status) {
    switch (status) {
    case Registry.STATUS.NOT_PAID:
      return 'NÃ£o pago'
    case Registry.STATUS.PAID:
      return 'Pago'
    case Registry.STATUS.PAID_FINANCIAL_CASH:
      return 'Depositar em poup'
    case Registry.STATUS.RESERVERD_SAVINGS:
      return 'Depositado em poup'
    default:
      return ''
    }
  }
}
