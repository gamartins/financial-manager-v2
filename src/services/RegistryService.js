const key = 'fmv2:registry'

const get = dateQuery => {
  const dataList = JSON.parse(localStorage.getItem(`${key}${dateQuery}`))

  if (!dataList) {
    localStorage.setItem(`${key}${dateQuery}`, JSON.stringify([]))
    return []
  }

  return dataList
}

const save = (dataList, dateQuery) => {
  localStorage.setItem(`${key}${dateQuery}`, JSON.stringify(dataList))
}

export default {
  get,
  save,
}
