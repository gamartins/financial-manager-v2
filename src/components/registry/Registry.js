import React, { Component } from 'react'

import addMonths from 'date-fns/addMonths'

import RegistryForm from './RegistryForm'
import RegistryList from './RegistryList'
import DatePagination from '../date-pagination/DatePagination'
import RegistryService from '../../services/RegistryService'

export default class Registry extends Component {

  static TYPE = {
    PROVENTO: 0,
    DESCONTO: 1,
  }

  constructor(props) {
    super(props)

    this.state = {
      date: new Date(),
      dataList: [],
      formValues: {
        id: null,
        name: '',
        price: 0,
        type: Registry.TYPE.DESCONTO,
        category: ''
      }
    }
  }

  componentDidMount() {
    this.getAll()
  }

  getAll() {
    const dateQuery = `${this.state.date.getFullYear()}${this.state.date.getMonth()}`

    const dataList = RegistryService.get(dateQuery)
    
    this.setState({ dataList })
  }

  getNext() {
    this.setState(
      (state) => ({ date: addMonths(state.date, 1) }),
      () => this.getAll())
  }

  getPrevious() {
    this.setState(
      (state) => ({ date: addMonths(state.date, -1) }),
      () => this.getAll())
  }

  setFormValue(id) {
    const index = this.state.dataList.findIndex(item => item.id === id)

    if (index === -1) {
      return
    }

    const newFormValues = { ...this.state.dataList[index] }
    
    this.setState({ formValues: newFormValues })
  }

  resetFormValue() {
    this.setState({
      formValues: {
        id: null,
        name: '',
        price: 0,
        type: Registry.TYPE.DESCONTO,
        category: '',
      }
    })
  }

  save(values) {
    return values.id ? this.edit(values) : this.add(values)
  }

  add(values) {
    const dateQuery = `${this.state.date.getFullYear()}${this.state.date.getMonth()}`
    const dataList = [ ...this.state.dataList ]
    dataList.push({
      id: Math.ceil((Math.random() * 100000)),
      name: values.name,
      price: values.price,
      type: parseInt(values.type),
      category: parseInt(values.category),
    })

    RegistryService.save(dataList, dateQuery)
    this.setState({ dataList: dataList })
    this.resetFormValue()

    return Promise.resolve()
  }

  edit(values) {
    const dateQuery = `${this.state.date.getFullYear()}${this.state.date.getMonth()}`
    const index = this.state.dataList.findIndex(item => item.id === values.id)
    const dataList = [ ...this.state.dataList ]
    dataList[index] = {
      id: values.id,
      name: values.name,
      price: values.price,
      type: parseInt(values.type),
      category: parseInt(values.category),
    }

    RegistryService.save(dataList, dateQuery)
    this.setState({ dataList: dataList })
    this.resetFormValue()

    return Promise.resolve()
  }

  remove(id) {
    const index = this.state.dataList.findIndex(item => item.id === id)

    if (index !== -1) {
      this.state.dataList.splice(index, 1)
      this.setState({ dataList: this.state.dataList })
      this.resetFormValue()
    }    
  }

  render() {
    return (
      <div>
        <RegistryForm
          initialValues={this.state.formValues}
          saveItem={values => this.save(values)} />
        <DatePagination
          date={this.state.date}
          format={'MMM'}
          previous={() => this.getPrevious()}
          next={() => this.getNext()}/>
        <RegistryList
          dataList={this.state.dataList}
          removeItem={id => this.remove(id)}
          editItem={item => this.setFormValue(item)} />
      </div>
    )
  }
}
