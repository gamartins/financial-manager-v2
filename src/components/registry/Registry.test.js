import React from 'react'
import Registry from './Registry'
import Enzyme, { mount, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

jest.mock('../../services/RegistryService');
import RegistryService from '../../services/RegistryService';

Enzyme.configure({ adapter: new Adapter() });

RegistryService.get.mockImplementation(() => [
    {
        id: 1,
        name: 'Registry 01',
        price: 2000,
        type: Registry.TYPE.PROVENTO,
        category: 1
    },
    {
        id: 2,
        name: 'Registry 02',
        price: 69,
        type: Registry.TYPE.DESCONTO,
        category: 1
    },
])

describe('<Registry />' ,() => {
    test('should be created', () => {
        const wrapper = mount(<Registry />)
        expect(wrapper).toMatchSnapshot()
    })

    test('should get registry list on componentDidMount', () => {
        const wrapper = shallow(<Registry />)
        expect(wrapper.instance().state.dataList).toHaveLength(2)
    })

    test('should find and set form value with item selected', () => {
        const wrapper = shallow(<Registry />)
        wrapper.instance().setFormValue(2)

        expect(wrapper.instance().state.formValues).toEqual({
            id: 2,
            name: 'Registry 02',
            price: 69,
            type: Registry.TYPE.DESCONTO,
            category: 1
        })
    })

    test('should call add when registry dont exists', () => {
        const wrapper = shallow(<Registry />)
        const addMock = wrapper.instance().add = jest.fn()
        wrapper.instance().save({
            id: null,
            name: 'Registry 02',
            price: 69,
            type: Registry.TYPE.DESCONTO,
            category: 1
        })

        expect(addMock).toHaveBeenCalled()
    })

    test('should call edit when registry already exists', () => {
        const wrapper = shallow(<Registry />)
        const editMock = wrapper.instance().edit = jest.fn()
        wrapper.instance().save({
            id: 2,
            name: 'Registry 02',
            price: 69,
            type: Registry.TYPE.DESCONTO,
            category: 1
        })

        expect(editMock).toHaveBeenCalled()
    })

    test('should remove one element when call remove', () => {
        const wrapper = mount(<Registry />)

        expect(wrapper.instance().state.dataList).toHaveLength(2)

        wrapper.instance().remove(2)

        expect(wrapper.instance().state.dataList).toHaveLength(1)
    })
})