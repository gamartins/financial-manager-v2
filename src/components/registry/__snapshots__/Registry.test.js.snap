// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Registry /> should be created 1`] = `ReactWrapper {}`;
